Exception when structural plasticity and multiple threads are used

sdiazpier opened the pull request:
This PR addresses issue #290 by throwing a kernel exception when structural plasticity is enabled and multiple threads have been set for simulation.

sdiazpier commented:
I suggest @heplesser, @jakobj and @sanjayankur31 as potential reviewers.

jakobj commented on nestkernel/sp_manager.cpp:
```diff
@@ -789,4 +789,25 @@ nest::SPManager::global_shuffle( std::vector< index >& v, size_t n )
+void
+nest::SPManager::enable_structural_plasticity()
+{
+  if ( kernel().vp_manager.get_num_threads() > 1 )
```
please add curly braces around the statement

jakobj commented on nestkernel/vp_manager.cpp:
```diff
@@ -180,6 +180,10 @@ nest::VPManager::get_status( DictionaryDatum& d )
 {
+  if ( kernel().sp_manager.is_structural_plasticity_enabled()
+    && ( n_threads > 1 ) )
+    throw KernelException(
```
(see above)

jakobj commented on pynest/nest/tests/test_sp/test_enable_multithread.py:
```diff
@@ -0,0 +1,73 @@
+        with self.assertRaises(nest.NESTError):
+            nest.SetKernelStatus(
+                {
+                    'resolution': 0.1,
```
why do you need to change resolution here?

heplesser commented:
@sdiazpier There are some small code-formatting problems and the test for issue #578 now seems to work no longer:
```
 Weight Recorder Single Threaded ... ok
  ERROR: test_targets (nest.tests.test_sp.test_issue_578_sp.TestIssue578)
      raise _kernel.NESTError(encode(errorstring))
  NESTError: Structural plasticity can not be used with multiple threads in EnableStructuralPlasticity: 
  FAILED (SKIP=11, errors=1)
```
That test requires two threads (it is run as a single MPI process):
```python
        nest.SetKernelStatus(
            {
                'resolution': 0.1,
                'total_num_virtual_procs': 2
            }
        )
```
So somehow that test and this PR seem orthogonal to each other?

heplesser commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -54,3 +63,10 @@ def suite():
+            call(["mpiexec", "-n", "2", "python", "test_sp/test_issue_578_sp.py"])
+        except:
+            print sys.exc_info()[0]
+            print "Test call with MPI ended in error"
```
`print` requires `()` to be Python3 compatible.

And shouldn't this exception propagate so that the test suite knows that the test failed? Currently, failing MPI will be registered as a working test.

heplesser commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -29,7 +29,17 @@
+    from mpi4py import MPI
+except ImportError:
+    # Test without MPI
+    print "Tests without MPI"
```
`print` needs `()` in Python3.

heplesser commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -29,7 +29,17 @@
+else:
+    # Test with MPI
+    mpi_test = 1
+    print "Testing with MPI"
```
`print` needs `()` in Python3.

heplesser commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -29,7 +29,17 @@
+except ImportError:
+    # Test without MPI
+    print "Tests without MPI"
+    mpi_test = 0
```
If MPI is not available, post a skipping message.

heplesser commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -54,3 +63,10 @@ def suite():
+    # MPI tests
+    if mpi_test:
+        try:
+            call(["mpiexec", "-n", "2", "python", "test_sp/test_issue_578_sp.py"])
```
It would be best to pick up the MPI-invocation rule from `.nestrc`.

sdiazpier commented:
Hi dear @heplesser, I have addressed most of your comments. I am not very acquainted with the format of the .nestrc file so I read the command in the most generic way I could with the knowledge that I have from it. Maybe I am not considering certain aspects of the format so If you have a better way or suggestion to make it easier/more robust please let me know.

sdiazpier commented:
Dear @jougs, could you please take a look at this PR and let me know your opinion on handling the nestrc file to obtain the mpi execution command?

jougs commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -43,11 +69,27 @@ def suite():
+            print ("MPI test command: " + line)
+            nestrcf.close()
+            return line
+
```
There's no need to parse and modify the line from `.nestrc` manually here as this is anyway a `SLI` function, which gives you the invocation line for a given number of processes, an executable and a script file. I suggest to replace the function signature by
```
def getMPITestCommand(np, script):
```
and the function body by something like this:
```
return nest.sli_func("mpirun", np, "python", script)
```

This is much safer, as it works for all definitions of `mpirun` in the `.nestrc` file.

jougs commented on pynest/nest/tests/test_sp/test_all.py:
```diff
@@ -28,12 +28,38 @@
+            command = getMPITestCommand()
+            command = command.replace(
+                "  scriptfile", "test_sp/mpitest_issue_578_sp.py")
+            print ("Executing test with command: " + command)
```
Given the changes to the `getMPITestCommand` function detailed below, you can replace lines 53-55 by just
```
command = getMPITestCommand(2, "test_sp/mpitest_issue_578_sp.py")
```

sdiazpier commented:
Hi dear @jougs I have made the changes you suggested. Could you please take a look at it and let me know if anything is missing?

sdiazpier commented:
Dear @jougs I implemented the changes we talked about and added the check for GSL in the test. Please let me know if anything else is needed.
